hooks/useRbfTransactionData/helpers.ts

```diff
@@ -0,0 +1,121 @@
+  rbfTxSummary?: {
+    currentFee: number;
+    currentFeeRate: number;
+    minimumRbfFee: number;
+    minimumRbfFeeRate: number;
+  };
+  rbfRecommendedFees?: RbfRecommendedFees;
+  mempoolFees?: RecommendedFeeResponse;
+  isLoading?: boolean;
+  errorCode?: 'SOMETHING_WENT_WRONG';
+};
+
+export const isBtcTransaction = (
+  transaction: BtcTransactionData | StxTransactionData,
+): transaction is BtcTransactionData => transaction?.txType === 'bitcoin';
+
+export const constructRecommendedFees = (
+  lowerName: keyof RbfRecommendedFees,
+  lowerFeeRate: number,
+  higherName: keyof RbfRecommendedFees,
+  higherFeeRate: number,
+  stxAvailableBalance: string,
+): RbfRecommendedFees => {
+  const bigNumLowerFee = BigNumber(lowerFeeRate);
+  const bigNumHigherFee = BigNumber(higherFeeRate);
+
+  return {
+    [lowerName]: {
+      enoughFunds: bigNumLowerFee.lte(BigNumber(stxAvailableBalance)),
+      feeRate: microstacksToStx(bigNumLowerFee).toNumber(),
+      fee: microstacksToStx(bigNumLowerFee).toNumber(),
+    },
+    [higherName]: {
+      enoughFunds: bigNumHigherFee.lte(BigNumber(stxAvailableBalance)),
+      feeRate: microstacksToStx(bigNumHigherFee).toNumber(),
+      fee: microstacksToStx(bigNumHigherFee).toNumber(),
+    },
+  };
+};
+
+export const sortFees = (fees: RbfRecommendedFees) =>
+  Object.fromEntries(
+    Object.entries(fees).sort((a, b) => {
+      const priorityOrder = ['highest', 'higher', 'high', 'medium'];
+      return priorityOrder.indexOf(a[0]) - priorityOrder.indexOf(b[0]);
+    }),
+  );
+
+export const calculateStxData = async (
+  transaction: StxTransactionData,
+  btcNetwork: SettingsNetwork,
+  stacksNetwork: StacksNetwork,
+  appInfo: AppInfo | null,
+  stxAvailableBalance: string,
+): Promise<RbfData> => {
+  const { fee } = transaction;
+  const txRaw: string = await getRawTransaction(transaction.txid, btcNetwork);
+  const unsignedTx: StacksTransaction = deserializeTransaction(txRaw);
+
+  const [slow, medium, high] = await estimateTransaction(unsignedTx.payload, undefined, stacksNetwork);
+
+  let feePresets: RbfRecommendedFees = {};
+  let mediumFee = medium.fee;
+  let highFee = high.fee;
+  const higherFee = fee.multipliedBy(1.25).toNumber();
+  const highestFee = fee.multipliedBy(1.5).toNumber();
+
+  if (appInfo?.thresholdHighStacksFee) {
+    if (high.fee > appInfo.thresholdHighStacksFee) {
+      // adding a fee cap
+      highFee = appInfo.thresholdHighStacksFee * 1.5;
+      mediumFee = appInfo.thresholdHighStacksFee;
+    }
+  }
+
+  let minimumFee = fee.multipliedBy(1.25).toNumber();
+  if (!Number.isSafeInteger(minimumFee)) {
+    // round up the fee to the nearest integer
+    minimumFee = Math.ceil(minimumFee);
+  }
+
+  if (fee.lt(BigNumber(mediumFee))) {
+    feePresets = constructRecommendedFees('medium', mediumFee, 'high', highFee, stxAvailableBalance);
+  } else {
+    feePresets = constructRecommendedFees('higher', higherFee, 'highest', highestFee, stxAvailableBalance);
+  }
```

Comment:
this is the business logic here that we need to unit test in helpers.test.ts

the unit tests should serve two purposes: 
- formalise the business logic intention, uncovering edge cases in the process
- prevent regressions in future when we need to alter the business logic, or extend functionality

how you do it will be down to code style, but if it were me, I would extract L78-102 into a separate pure function. then you won't need to mock the functions with side effects like estimateTransaction, making the test file lighter and less coupled to code which we are not trying to test here

Reply:
yeah, in fact this exact thing is what is currently covered in `helpers.test.ts` in two assertions, `should return the fee estimation fetched from the Stacks function` and `should return the fee estimation based on the current fee`. do you think we should write more tests for this logic?

Reply:
yup, but do you cover the threshold, or `fee.lt(BigNumber(mediumFee))` code paths? the unit tests should at the least cover the logic which was defined in the notion specs, and ideally more edge cases.

Reply:
yeah, the `fee.lt(BigNumber(mediumFee))` code paths were covered there, but I agree that it looked not that obvious, and not easy to read and update in the future.

my initial idea was to test the whole hook, but then I realized that it would be better to make the testing more concise and I like your idea about testing the pure function that is used only for calculation, so I just pushed an update to implement it.

Reply:
love it, thank you